crates/ruff_linter/src/rules/flake8_async/rules/long_sleep_not_forever.rs

```diff
@@ -0,0 +1,143 @@
+use ruff_python_semantic::Modules;
+use ruff_text_size::Ranged;
+
+use crate::checkers::ast::Checker;
+use crate::importer::ImportRequest;
+use crate::rules::flake8_async::helpers::AsyncModule;
+use crate::settings::types::PreviewMode;
+
+/// ## What it does
+/// Checks for uses of `trio.sleep()` or `anyio.sleep()` with a delay greater than 24 hours.
+///
+/// ## Why is this bad?
+/// Calling `sleep()` with a delay greater than 24 hours is usually intended
+/// to sleep indefinitely. Instead of using a large delay,
+/// `trio.sleep_forever()` or `anyio.sleep_forever()` better conveys the intent.
+///
+///
+/// ## Example
+/// ```python
+/// import trio
+///
+///
+/// async def func():
+///     await trio.sleep(86401)
+/// ```
+///
+/// Use instead:
+/// ```python
+/// import trio
+///
+///
+/// async def func():
+///     await trio.sleep_forever()
+/// ```
+#[violation]
+pub struct LongSleepNotForever {
+    module: AsyncModule,
+}
+
+impl Violation for LongSleepNotForever {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let Self { module } = self;
+        format!(
+            "`{module}.sleep()` with >24 hour interval should usually be `{module}.sleep_forever()`"
+        )
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        let Self { module } = self;
+        Some(format!("Replace with `{module}.sleep_forever()`"))
+    }
+}
+
+/// ASYNC116
+pub(crate) fn long_sleep_not_forever(checker: &mut Checker, call: &ExprCall) {
+    if !(checker.semantic().seen_module(Modules::TRIO)
+        || checker.semantic().seen_module(Modules::ANYIO))
+    {
+        return;
+    }
+
+    if call.arguments.len() != 1 {
+        return;
+    }
+
+    let Some(arg) = call.arguments.find_argument("seconds", 0) else {
+        return;
+    };
+
+    let Some(qualified_name) = checker
+        .semantic()
+        .resolve_qualified_name(call.func.as_ref())
+    else {
+        return;
+    };
+
+    if !matches!(qualified_name.segments(), ["trio" | "anyio", "sleep"]) {
+        return;
+    }
+
+    let Expr::NumberLiteral(ExprNumberLiteral { value, .. }) = arg else {
+        return;
+    };
+
+    // TODO(ekohilas): Replace with Duration::from_days(1).as_secs(); when available.
+    let one_day_in_secs = 60 * 60 * 24;
+    match value {
+        Number::Int(int_value) => {
+            let Some(int_value) = int_value.as_u64() else {
+                return;
+            };
+            if int_value <= one_day_in_secs {
+                return;
+            }
+        }
+        Number::Float(float_value) =>
+        {
+            #[allow(clippy::cast_precision_loss)]
+            if *float_value <= one_day_in_secs as f64 {
+                return;
+            }
+        }
+        Number::Complex { .. } => return,
+    }
+
+    let module = AsyncModule::try_from(&qualified_name).unwrap();
```

Comment:
Maybe we extract this simultaneously with the `!matches!(qualified_name.segments(), ["trio" | "anyio", "sleep"])` above, to make the `unwrap` unnecessary?

Reply:
I'm not sure what pattern you're suggesting

Reply:
Maybe something like

```rust
let module = match qualfied_name.segments() {
   ["trio", "sleep"] => AsyncModule::Trio,
   ["anyio", "sleep"] => AsyncModle::Anyio,
   _ => return;
};
```

or

```rust
let Some(module) = AsyncModule::try_from(&qualified_name) else { return; };
if !matches!(qualified_name.segments(), [_, "sleep"]) { return; };
```